bigquery/google/cloud/bigquery/_helpers.py

```diff
@@ -228,16 +230,280 @@ def __set__(self, instance, value):
         instance._udf_resources = tuple(value)
 
 
-def _build_udf_resources(resources):
+class AbstractQueryParameter(object):
+    """Base class for named / positional query parameters.
     """
-    :type resources: sequence of :class:`UDFResource`
-    :param resources: fields to be appended.
+    @classmethod
+    def from_api_repr(cls, resource):
+        """Factory: construct paramter from JSON resource.
+
+        :type resource: dict
+        :param resource: JSON mapping of parameter
+
+        :rtype: :class:`ScalarQueryParameter`
+        """
+        raise NotImplementedError
+
+    def to_api_repr(self):
+        """Construct JSON API representation for the parameter.
+
+        :rtype: dict
+        """
+        raise NotImplementedError
+
 
-    :rtype: mapping
-    :returns: a mapping describing userDefinedFunctionResources for the query.
+class ScalarQueryParameter(AbstractQueryParameter):
+    """Named / positional query parameters for scalar values.
+
+    :type name: str or None
+    :param name: Parameter name, used via ``@foo`` syntax.  If None, the
+                 paramter can only be addressed via position (``?``).
+
+    :type type_: str
+    :param type_: name of parameter type.  One of `'STRING'`, `'INT64'`,
+                  `'FLOAT64'`, `'BOOLEAN'`, `'TIMESTAMP'`, or `'DATE'`.
```

Comment:
Hmm, the [API docs for schema definition](https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#schema.fields.type) use `BOOLEAN`:  is `BOOL` the "standard SQL" alias (like `INT64` for `INTEGER`)?